src/blockchain.rs

```diff
@@ -39,20 +58,169 @@ impl Blockchain {
 
     /// Returns the latest `Block` as part of the longest chain
     pub fn get_latest_block_index(&self) -> Option<&BlockIndex> {
-        self.index.blocks.last()
+        match self.index.last_block_hash {
+            Some(last_block_hash) => self.index.block_map.get(&last_block_hash),
+            None => None,
+        }
+    }
+
+    pub fn get_genesis_block_index(&self) -> Option<&BlockIndex> {
+        match self.index.genesis_block_hash {
+            Some(genesis_block_hash) => self.index.block_map.get(&genesis_block_hash),
+            None => None,
+        }
     }
 
     /// Append `Block` to the index of `Blockchain`
     ///
     /// * `block` - `Block` appended to index
-    pub fn add_block(&mut self, block: Block) {
+    pub fn try_add_block(&mut self, block: Block) -> crate::Result<()> {
+        // TODO -- we want to lock writes to the blockchain while we're adding a block
+        // To do that, we should implement a RwLock around the `Blockchain` index to verify
+        // that there's only one thread that is writing to the chain at once time
+
+        if let Some((genesis_block_header, _)) = self.get_genesis_block_index() {
+            // check if our candidate blocks is greater than the genesis block
+            if self.genesis_block_check(&genesis_block_header, &block) {
+                // Add the block to the index, but not the longest chain just yet
+                self.add_block_index(&block);
+
+                // If our candidate block previous block_hash is equal to last_block_hash,
+                // then the candidate block is building on the longest chain
+                if &block.header().previous_block_hash() == &(self.index.last_block_hash.unwrap()) {
```

Comment:
I think before we get here we shoudl check if the block's parent_hash is in our hashmap